## __proteinSSpredictor__

This projects constitutes a final project for Data Science Bootcamp organised by Sages and aims at providing a deep-learning-based approach to predict secondary structure of the protein based on its primary sequence of the amino acids.

## __Table of contents__
1. Abstract
2. Source of raw sequences
3. Curation of the dataset
4. Prerequisites
5. Raw data processing
6. Model training
7. Where is the model making errors?
8. Summary
9. References

## __1. Abstract__

to do

## __2. Source of raw sequences__

Preparation of the updated dataset analogous to https://www.kaggle.com/alfrandom/protein-secondary-structure

Following steps from https://github.com/zyxue/pdb-secondary-structure the secondary structure data was downloaded by

`DATE_STAMP=$(date  +'%Y-%m-%d')
wget https://cdn.rcsb.org/etl/kabschSander/ss.txt.gz -O ${DATE_TAMP}-ss.txt.gz`

Further, data was transformed to csv file analogously to https://github.com/zyxue/pdb-secondary-structure using self-implemented method (see below for more details).

## __3. Curation of the sequence dataset__ 

Further, recent PISCES data with removed peptides with high sequence similarity was downloaded from http://dunbrack.fccc.edu/Guoli/pisces_download.php

Selected precompiled CulledPDB for further modeling was generated with following criteria:

- percentage identity cutoff: 30%
- the resolution cutoff:      2.0 Å
- the R-factor cutoff:        0.25
- date of generation:         2021/07/02
- the number of sequences:    10870


## __4. Prerequisites__

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt


## __4. Raw data processing__
At this point the raw sequences were converted to CSV file, translated to 3 letters representation, and combined with culledPDB list with reduced sequence identity to improve the model quality

In [9]:
from tools import process_raw_data

In [11]:
processing = process_raw_data.Sequences()

In [2]:
%time sequences_df = processing.process_raw_sequences('../data/2021-07-09-ss.txt.gz')

processed 477153 sequences
CPU times: user 10.9 s, sys: 427 ms, total: 11.3 s
Wall time: 11.8 s


In [12]:
#optional for faster processing:
sequences_df = pd.read_csv('../data/2021-07-09-ss.csv')


The set of raw sequences before processing with CulledPDB dataset contains 477154 and includes information about the sequence PDB ID code, protein chain and its primary sequence together with the secondary structure sequence

In [13]:
sequences_df.shape

(477154, 4)

In [14]:
sequences_df.head()

pdb_id chain                                           sequence  \
0   101M     A  MVLSEGEWQLVLHVWAKVEADVAGHGQDILIRLFKSHPETLEKFDR...   
1   102L     A  MNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPSLNAAAKSE...   
2   102M     A  MVLSEGEWQLVLHVWAKVEADVAGHGQDILIRLFKSHPETLEKFDR...   
3   103L     A  MNIFEMLRIDEGLRLKIYKDTEGYYTIGIGHLLTKSPSLNSLDAAK...   
4   103M     A  MVLSEGEWQLVLHVWAKVEADVAGHGQDILIRLFKSHPETLEKFDR...   

                                    secondary_struct  
0  CCCCHHHHHHHHHHHHHHGGGHHHHHHHHHHHHHHHCGGGGGGCTT...  
1  CCHHHHHHHHHCCEEEEEECTTSCEEEETTEEEESSSCTTTHHHHH...  
2  CCCCHHHHHHHHHHHHHHGGGHHHHHHHHHHHHHHHCGGGGGGCTT...  
3  CCHHHHHHHHHCCEEEEEECTTSCEEEETTEECCCCCCCCCHHHHH...  
4  CCCCHHHHHHHHHHHHHHGGGHHHHHHHHHHHHHHHCGGGGGGCTT...

In [15]:
pisces_df = processing.clean_pisces(pisces_file='../data/cullpdb_pc30_res2.0_R0.25_d2021_07_02_chains10870.gz')

In [16]:
pisces_df.head()

length source  resolution  R_value  R_free pdb_id chain
0     108   XRAY         2.0     0.21    0.25   1A1X     A
1     148   XRAY         1.8     0.19    0.24   1A3A     A
2     163   XRAY         1.8     0.21    0.30   1A73     A
3     226   XRAY         1.9     0.19    0.22   1A8L     A
4    1073   XRAY         1.8     0.19    1.00   1A9X     C

The culledPDB set containts 10870 items (PDB databases records filtered by the resolution, R-value and the 30% sequence identity). It contains information about protein PDB ID code, protein chain, method used for its determination, R-value, R-value free and the length of the sequence.

In [17]:
pisces_df.shape

(10870, 7)

In [18]:
combined_df = processing.combine_pdb_pisces(sequences_df=sequences_df, pisces_df=pisces_df)

In [19]:
combined_df.head()

pdb_id chain                                           sequence  \
0   1A1X     A  GSAGEDVGAPPDHLWVHQEGIYRDEYQRTWVAVVEEETSFLRARVQ...   
1   1A3A     A  MANLFKLGAENIFLGRKAATKEEAIRFAGEQLVKGGYVEPEYVQAM...   
2   1A73     A  MALTNAQILAVIDSWEETVGQFPVITHHVPLGGGLQGTLHCYEIPL...   
3   1A8L     A  MGLISDADKKVIKEEFFSKMVNPVKLIVFVRKDHCQYCDQLKQLVQ...   
4   1A9X     F  IKSALLVLEDGTQFHGRAIGATGSAVGEVVFNTSMTGYQEILTDPS...   

                                    secondary_struct  length source  \
0  CCCCCCCCCCCSEEEEEETTEEEETTSCEEEEEEEECSSCEEEEEE...     108   XRAY   
1  CCCCCCCCGGGEECSCCCSSHHHHHHHHHHHHHHTTSBCTHHHHHH...     148   XRAY   
2  CCCCHHHHHHHHHHHHHHHTTSCEEEEEEEEETTEEEEEEEEBCCS...     163   XRAY   
3  CCSSCHHHHHHHHHHTGGGCCSCEEEEEEECSSSCTTHHHHHHHHH...     226   XRAY   
4  CCEEEEEETTCCEEEEEECSCSEEEEEEEEEECCSSCHHHHHTCGG...     379   XRAY   

   resolution  R_value  R_free  
0         2.0     0.21    0.25  
1         1.8     0.19    0.24  
2         1.8     0.21    0.30  
3         1.9     0.19    0.22  
4         1.8     0.19    1.00

In [20]:
combined_df.shape

(10186, 9)

While intersecting raw sequences and culledPDB dataset by protein PDB ID codes and chains, we obtain slightly smaller set (10186) in then the original culledPDB (10870) due to some mismatches in the chain IDs

In [21]:
print(sorted(pisces_df.chain.unique()))
print(f'amount of unique chain IDs in PISCES dataset: {len(pisces_df.chain.unique())}')

['1', '2', '3', '4', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
amount of unique chain IDs in PISCES dataset: 30


In [22]:
print(sorted(sequences_df.chain.unique()))
print(f'amount of unique chain IDs in dataset of raw sequences: {len(sequences_df.chain.unique())}')

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z']
amount of unique chain IDs in dataset of raw sequences: 36


As the protein secondary structure can be represented in the extended form containing 8 secondary structure types (Q8) and simplified, containing only 3 secondary structure types (Q3), we create new column with the simplified version for modeling

Q8: Q3

__helices__ G, H, I: H

__sheet__ B, E: E

__coil__ T, S, C: C

Additionally, we get rid of the remaining columns and keep just PDB ID code, sequence, extended (Q8) secondary structure representation and simplified (Q3) representation as pdb_id, seq, sst_q8 and sst_q3, respectively:

In [23]:
df_for_modeling = processing.generate_seq_q8_q3_representation(combined_df)

In [24]:
df_for_modeling

pdb_id                                                seq  \
0       1A1X  GSAGEDVGAPPDHLWVHQEGIYRDEYQRTWVAVVEEETSFLRARVQ...   
1       1A3A  MANLFKLGAENIFLGRKAATKEEAIRFAGEQLVKGGYVEPEYVQAM...   
2       1A73  MALTNAQILAVIDSWEETVGQFPVITHHVPLGGGLQGTLHCYEIPL...   
3       1A8L  MGLISDADKKVIKEEFFSKMVNPVKLIVFVRKDHCQYCDQLKQLVQ...   
4       1A9X  IKSALLVLEDGTQFHGRAIGATGSAVGEVVFNTSMTGYQEILTDPS...   
...      ...                                                ...   
10181   7BU2  MSMIKSYAAKEAGGELEVYEYDPGELRPQDVEVQVDYCGICHSDLS...   
10182   7BVD  MQTTANHSSRSTQTGTRAHGAALAETTSREDFRALATEHRVVPVIR...   
10183   7BYW  MRGSHHHHHHSGQRMGMVIGIKPEHIDEYKRLHAAVWPAVLARLAE...   
10184   7ODC  MSSFTKDEFDCHILDEGFTAKDILDQKINEVSSSDDKDAFYVADLG...   
10185   8ABP  ENLKLGFLVKQPEEPWFQTEWKFADKAGKDLGFEVIKIAVPDGEKT...   

                                                  sst_q8  \
0      CCCCCCCCCCCSEEEEEETTEEEETTSCEEEEEEEECSSCEEEEEE...   
1      CCCCCCCCGGGEECSCCCSSHHHHHHHHHHHHHHTTSBCTHHHHHH...   
2      CCCCHHHHHHHHHHHHHHHTTSCEEEEEEEEETTEEEEEEEEBCCS...   
3      CCSSCHHHHHHHHHHTGGGCCSCEEEEEEECSSSCTTHHHHHHHHH...   
4      CCEEEEEETTCCEEEEEECSCSEEEEEEEEEECCSSCHHHHHTCGG...   
...                                                  ...   
10181  CCCEEEEEESSTTCCEEEEEECCCSCCTTEEEEEEEEEEECHHHHH...   
10182  CCCCCCCCCCCCCCCCCCCCCCTTCCCCHHHHHHHHTTCSEEEEEE...   
10183  CCCCCCCCCCCCCEEEEEEEECGGGHHHHHHHHTTCCHHHHHHHHH...   
10184  CCEEEETTEEEEEECSSCCHHHHHHHHHCCCCCCCCCCCEEEEEHH...   
10185  CCEEEEEEESCTTSHHHHHHHHHHHHHHHHHTEEEEEEECCSHHHH...   

                                                  sst_q3  
0      CCCCCCCCCCCCEEEEEECCEEEECCCCEEEEEEEECCCCEEEEEE...  
1      CCCCCCCCHHHEECCCCCCCHHHHHHHHHHHHHHCCCECCHHHHHH...  
2      CCCCHHHHHHHHHHHHHHHCCCCEEEEEEEEECCEEEEEEEEECCC...  
3      CCCCCHHHHHHHHHHCHHHCCCCEEEEEEECCCCCCCHHHHHHHHH...  
4      CCEEEEEECCCCEEEEEECCCCEEEEEEEEEECCCCCHHHHHCCHH...  
...                                                  ...  
10181  CCCEEEEEECCCCCCEEEEEECCCCCCCCEEEEEEEEEEECHHHHH...  
10182  CCCCCCCCCCCCCCCCCCCCCCCCCCCCHHHHHHHHCCCCEEEEEE...  
10183  CCCCCCCCCCCCCEEEEEEEECHHHHHHHHHHHCCCCHHHHHHHHH...  
10184  CCEEEECCEEEEEECCCCCHHHHHHHHHCCCCCCCCCCCEEEEEHH...  
10185  CCEEEEEEECCCCCHHHHHHHHHHHHHHHHHCEEEEEEECCCHHHH...  

[10186 rows x 4 columns]

In [69]:
test_df = pd.DataFrame()
test_df['seq_len'] = [len(x) for x in df_for_modeling.seq]
test_df['sst_q8_len'] = [len(x) for x in df_for_modeling.sst_q8]
test_df['sst_q3_len'] = [len(x) for x in df_for_modeling.sst_q3]

# [len(x) for x in df_for_modeling.seq]

In [75]:
test_df[test_df.sst_q3_len != test_df.sst_q8_len]

Empty DataFrame
Columns: [seq_len, sst_q8_len, sst_q3_len]
Index: []

In [34]:
len(df_for_modeling.iloc[0].seq)

108